curvefs/src/client/s3/client_s3_adaptor.cpp

```diff
@@ -0,0 +1,580 @@
+            result.push_back(tmp);
+        /*
+             ----------     old
+               ------       new
+        */       
+        } else {
+            tmp.set_chunkid(old.chunkid());
+            tmp.set_version(old.version());
+            tmp.set_offset(old.offset());
+            tmp.set_len(newChunk.offset() - old.offset());
+            tmp.set_size(newChunk.offset() - old.offset());
+            result.push_back(tmp);
+            tmp.set_chunkid(old.chunkid());
+            tmp.set_version(old.version());
+            tmp.set_offset(newChunk.offset() + newChunk.len());
+            tmp.set_len(old.offset() + old.len()
+            - newChunk.offset() - newChunk.len());
+            tmp.set_size(old.offset() + old.len()
+            - newChunk.offset() - newChunk.len());
+            result.push_back(tmp);
+        }
+    /*
+                  -----     old
+               ----------   new
+    */    
+    } else if (newChunk.offset() <= old.offset()
+      && newChunk.offset() + newChunk.len() >= old.offset() + old.len()) {
+        return result;
+    /*
+                  --------  old
+               -------      new
+    */ 
+    } else {
+        tmp.set_chunkid(old.chunkid());
+        tmp.set_version(old.version());
+        tmp.set_offset(newChunk.offset()+ newChunk.len());
+        tmp.set_len(old.offset() + old.len()
+        - newChunk.offset() - newChunk.len());
+        tmp.set_size(old.offset() + old.len()
+        - newChunk.offset() - newChunk.len());
+        result.push_back(tmp);
+    }
+
+    return result;
+}
+
+std::vector<S3ChunkInfo> S3ClientAdaptor::GetReadChunks(Inode *inode) {
+    S3ChunkInfo tmp, chunkTmp;
+    std::vector<S3ChunkInfo> chunks;
+    const S3ChunkInfoList& s3ChunkInfoList = inode->s3chunkinfolist();
+
+    for (int i = 0; i < s3ChunkInfoList.s3chunks_size(); ++i) {
+        tmp = s3ChunkInfoList.s3chunks(i);
+        std::vector<S3ChunkInfo> addChunks;
+        std::vector<int> waitingDel;
+        for (uint32_t j = 0; j < chunks.size(); j++) {
+            chunkTmp = chunks[j];
+            // overlap, must cut old chunk
+            if ((tmp.offset() < (chunkTmp.offset() + chunkTmp.len())) &&
+                (chunkTmp.offset() < (tmp.offset() + tmp.len()))) {
+                addChunks = CutOverLapChunks(tmp, chunkTmp);
+                waitingDel.push_back(j);
+            }
+        }
+
+        std::vector<int>::iterator iter = waitingDel.begin();
+        for (; iter != waitingDel.end(); iter++) {
+            chunks.erase(chunks.begin() + *iter);
+        }
+        std::vector<S3ChunkInfo>::iterator chunkIter = addChunks.begin();
+        for (; chunkIter != addChunks.end(); chunkIter++) {
+            chunks.push_back(*chunkIter);
+        }
+        chunks.push_back(tmp);
+    }
+
+    return chunks;
+}
+
+std::vector<S3ChunkInfo> S3ClientAdaptor::SortByOffset(
+                             std::vector<S3ChunkInfo> chunks) {
+    int i, j;
```

Comment:
use std::sort and pass a comparator

Reply:
std::sort(chunks.begin(), chunks.end(), [] (S3ChunkInfo a, S3ChunkInfo b) {
        return a.offset < b.offset;
    });